### PR TITLE
Partial reading for `JsonReader.Either`

### PR DESCRIPTION
**Pull request by bastbu**
This PR introduces partial JSON reading support for all `JsonReader.Either` overloads.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -236,11 +236,7 @@ then the `Incomplete` property of the returned `JsonReadResult<T>` will be
 `true`. This is a signal to the caller that it must load the buffer with more
 of the JSON source text to resume reading.
 
-Presently, the following readers do not support partial reading:
-
-- `JsonReader.Either`
-
-Nevertheless, they can be composed with `JsonReader.Buffer` to load enough
+All readers can be composed with `JsonReader.Buffer` to load enough
 data into the buffer that neither can fail due to partial JSON.
```
Comment (atifaziz) [inline README.md]:
```suggestion
data into the buffer so that none can fail due to partial JSON.
```

File: src/JsonReader.cs [hunk 2]
```diff
@@ -484,26 +484,64 @@ public static IJsonReader<object> AsObject<T>(this IJsonReader<T> reader)
...
 
+    enum EitherSideState { Incomplete, Error }
+
     public static IJsonReader<T> Either<T>(IJsonReader<T> reader1, IJsonReader<T> reader2,
                                            string? errorMessage) =>
         Create((ref Utf8JsonReader rdr) =>
         {
-            var irdr = rdr;
-            switch (reader1.TryRead(ref irdr))
+            var (leftStack, rightStack, leftState, rightState) = rdr.ResumeOrDefault<(Stack<object>?, Stack<object>?, EitherSideState, EitherSideState)>();
```
Comment (atifaziz) [inline src/JsonReader.cs]: nit: thoughts on giving this more structure?

```c#
var ((leftStack, leftState), (rightStack, rightState)) =
    rdr.ResumeOrDefault<((Stack<object>?, EitherSideState), (Stack<object>?, EitherSideState))>();
```

or:

```c#
var (left, right) =
    rdr.ResumeOrDefault<((Stack<object>? Stack, EitherSideState State),
                         (Stack<object>? Stack, EitherSideState State))>();
```

The latter can also make the suspension look simpler:

```c#
return rdr.Suspend((left, right));
```

**Review by atifaziz — APPROVED** [review]
Brilliant and thanks!

File: src/JsonReader.cs (hunk 2 above)
Comment (bastbu) [reply src/JsonReader.cs]: I like it, applied your second suggestion with 0f3df136609273047c601e9c54b742692f2efda8

**Pull request merged**